webapp/src/marked.tsx

```diff
@@ -0,0 +1,117 @@
+
+import * as React from "react";
+import * as ReactDOM from "react-dom";
+import * as data from "./data";
+import * as marked from "marked";
+
+type ISettingsProps = pxt.editor.ISettingsProps;
+
+interface MarkedContentProps extends ISettingsProps {
+    markdown: string;
+}
+
+interface MarkedContentState {
+}
+
+export class MarkedContent extends data.Component<MarkedContentProps, MarkedContentState> {
+
+    // Local cache for images, cleared when we create a new project.
+    // Stores code => data-uri image of decompiled result
+    private static blockSnippetCache: pxt.Map<string> = {};
+    public static clearBlockSnippetCache() {
+        this.blockSnippetCache = {};
+    }
+
+    private getBuiltinMacros() {
+        const params: pxt.Map<string> = {};
+        const theme = pxt.appTarget.appTheme;
+        if (theme.boardName)
+            params["boardname"] = pxt.Util.htmlEscape(theme.boardName);
+        if (theme.boardNickname)
+            params["boardnickname"] = pxt.Util.htmlEscape(theme.boardNickname);
+        if (theme.driveDisplayName)
+            params["drivename"] = pxt.Util.htmlEscape(theme.driveDisplayName);
+        if (theme.homeUrl)
+            params["homeurl"] = pxt.Util.htmlEscape(theme.homeUrl);
+        params["targetid"] = theme.id || "???";
+        params["targetname"] = theme.name || "Microsoft MakeCode";
+        params["targetlogo"] = theme.docsLogo ? `<img aria-hidden="true" role="presentation" class="ui mini image" src="${theme.docsLogo}" />` : "";
+        return params;
+    }
+
+    private renderSnippets(content: HTMLElement) {
+        const { parent } = this.props;
+
+        pxt.Util.toArray(content.querySelectorAll(`.lang-blocks`))
+            .forEach((langBlock: HTMLElement) => {
+                const code = langBlock.innerText;
+                langBlock.innerHTML = `<div class="ui segment raised loading"></div>`;
+                if (MarkedContent.blockSnippetCache[code]) {
+                    // Use cache
+                    langBlock.innerHTML = `<div class="ui segment raised">
+                                <img src="${MarkedContent.blockSnippetCache[code]}" /></div>`;
+                } else {
+                    parent.renderBlocksAsync({
+                        action: "renderblocks", ts: code
+                    } as pxt.editor.EditorMessageRenderBlocksRequest)
+                        .done((xml) => {
+                            if (xml) {
+                                langBlock.innerHTML = `<div class="ui segment raised">
+                                    <img src="${pxt.Util.htmlEscape(xml)}" /></div>`;
+                                MarkedContent.blockSnippetCache[code] = xml;
```

Comment:
escape xml before caching. It would be safer to contruct DOM nodes that rely on innerHTML.